### ISSUE TITLE
Strange Redirection Problems with xonsh in mlterm

### ISSUE THREAD

**Issue by xee5ch** [xonsh/xonsh#741]
I cannot tell if this is something strange I have done with my mlterm configuration, which is minimal, or my new xonsh configuration. I usually use emacs as a term multiplexer these days, and mlterm was a hold-over for complex script languages I read and sometimes use. I am unable to delete a file when prompted, it immediatetly throws an error when prompted for user input.

```
me@computer ~ $ mlterm -v
mlterm version 3.7.0
me@computer ~ $ import xonsh
me@computer ~ $ xonsh.__version__
'0.2.6'
me@computer ~ $ import prompt_toolkit
me@computer ~ $ SHELL_TYPE=prompt_toolkit
me@computer ~ $ prompt_toolkit.__version__
'0.60'
me@computer ~ $ rm ~/Code/conf.py
/usr/bin/rm: remove regular file '/home/me/Code/conf.py'?
*[1] stopped: rm /home/me/Code/conf.py & (10123)
TypeError: unorderable types: NoneType() > int()

For the last line, if I redirect output with xclip to speed up copy-paste for this ticket, redirection magically is fine and I can actually answer yes or no.


me@computer ~ $ rm ~/Code/conf.py | xclip -sel clip
/usr/bin/rm: remove regular file '/home/me/Code/conf.py'?
```

My dotfiles are in my repo. I just created the most minimal xonsh possible from my minimal needs with `xonfig`.

```
{
 "foreign_shells": [
  {
   "login": true,
   "shell": "bash"
  },
  {
   "shell": "bash"
  }
 ]
}
```

Did I wrap two login shells into each other or something basic I forgot? I have tried running mlterm straight with the default bash shell, but also tried bypassing this and running just xonsh with `mlterm -e xonsh` and I get the same behavior on all attempts.

**Comment by gforsyth** [comment]
Hey @ajstein 
In your `.bashrc` or other `rc` files, do you have an alias for `rm` to `rm -I`?  This might be related to #727

**Comment by ajstein** [comment]
As a matter of fact, yes.  I had not even recalled.

https://github.com/ajstein/dotfiles/blob/personal_linux/.bash_aliases#L3

```
# general aliases
alias ls='ls --color=auto'
alias rm='rm -iv '

```

Does this mean I should proceed with the `andor` branch?

**Comment by gforsyth** [comment]
Give `andor` a shot and let us know what happens -- it should fix the `TypeError: unorderable types: NoneType() > int()` error, but I don't think `rm` will behave as expected.  

What linux distro are you using?

**Comment by gforsyth** [comment]
Ahh, ok, you're on Arch.  Me too.

**Comment by ajstein** [comment]
Alright.  I can give it a shot.

OS: Arch Linux
```
me@computer ~ $ uname -a 
Linux littlewad 4.4.5-1-ARCH #1 SMP PREEMPT Thu Mar 10 07:38:19 CET 2016 x86_64 GNU/Linux
```
Full list of [packages listed in this Gist](https://gist.github.com/ajstein/bee77837197a5aa04e51).

**Comment by ajstein** [comment]
As you guessed, no dice.  I quickly built with the same deps, but pulled down the `andor` branch using this custom AUR PKGBUILD file for quick testing.

https://gist.github.com/ajstein/31064e8d2807939eca61

```
me@computer ~ $ rm ~/Code/conf.py
/usr/bin/rm: remove regular file '/home/me/Code/conf.py'?
*[1] stopped: rm /home/me/Code/conf.py & (15080)
```

`mlterm -e xonsh` seems to have better syntax highlighting (maybe I called `import prompt_toolkit; SHELL_TYPE=prompt_toolkit` the right way this time, but i still get the above error.  Back to the master branch with AUR@xonsh-git for me I guess.

**Comment by gforsyth** [comment]
Ok -- I had just added the `rm = rm -I` alias recently, but didn't think to check if it worked on previous releases.  It does!  
Bisection points to 0a46ebf54 as the culprit, which introduced `settitle()`

If I comment out https://github.com/scopatz/xonsh/blob/49e4f18debaef4529734d5252f1ad79be8b67371/xonsh/built_ins.py#L640-L647 then `rm -I` works as expected.  

This removes the title from the window, as you'd expect.  I'm trying to work out a fix that preserves functionality.

**Comment by ajstein** [comment]
Well I am a longstanding novice in Python and a shy F/OSS contributor over the years and these seems like a fun project.  I wanted to know how shells work.  Is there anything I can do that is more than navel gazing?  I will look at the code myself.

**Comment by gforsyth** [comment]
By all means!  It helps to have someone else with this bug, too.  

I've narrowed it down to the function `settitle()` in `base_shell.py` but I cannot fathom why it is behaving the way that it is.  If you have any ideas or spot a fix, feel free to let me know or to put in a PR.

**Comment by adqm** [comment]
Welcome, @ajstein!

This is a really interesting issue!  And I can't reproduce it, even with those aliases.

Are you running the current master?  I very recently (in #702) modified `settitle` in a way that may fix this.  Do you mind testing that out and letting me know if that solves it?

**Comment by gforsyth** [comment]
Hey @adqm -- no kidding.  I thought I was going crazy.  What distro are you running?  And what version of `coreutils`?  I don't know why, but that seems to be one of the main points of difference between Arch and other distros.

**Comment by adqm** [comment]
It's a mystery!  How exciting...

I'm on Trisquel 7.0, with coreutils 8.21.

**Comment by ajstein** [comment]
Well, unless otherwise stated, I am on Arch and I use the AUR system to build certain unofficial packages from source. [Xonsh-git](https://aur.archlinux.org/packages/xonsh-git) is no exception.  I tried building my own build with `andor` above, but before and after that request I built against master.  This is indicated by the Arch pkg manager as:

```
[me@computer dotfiles]$ pacman -Qi xonsh-git
Name            : xonsh-git
Version         : 0.2.6.r250.g49e4f18-1
Description     : A Python-ish, BASHwards-compatible shell
Architecture    : any
URL             : http://github.com/scopatz/xonsh
Licenses        : FreeBSD
Groups          : None
Provides        : None
Depends On      : python  python-ply
Optional Deps   : None
Required By     : None
Optional For    : None
Conflicts With  : xonsh
Replaces        : None
Installed Size  : 1822.00 KiB
Packager        : Me <blah@blah.blah>
Build Date      : Wed 16 Mar 2016 10:18:21 PM AST
Install Date    : Wed 16 Mar 2016 10:18:25 PM AST
Install Reason  : Explicitly installed
Install Script  : Yes
Validated By    : None
```

This is [the recipe](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=xonsh-git), unlike my custom one above, that will always build from master.

**Comment by adqm** [comment]
Thanks, @ajstein.  The only question, then, is whether that install happened within the last ~15 hours (since the PR in question was only merged in _very_ recently).

**Comment by ajstein** [comment]
It's also possible I misunderstood, but I am pretty sure it is building the most recent copy.  Let me double-check the Arch docs.  I thought @gforsyth would correct me if I am wrong as he is also an Arch user.

I am re-reading the packaging guidelines for AUR & VCS before I nod off for the day.

**Comment by gforsyth** [comment]
I believe that the `xonsh-git` AUR build will always pull in the latest version of master at build time.

**Comment by adqm** [comment]
Yes, that looks right from the recipe.  My question is whether the version on @ajstein 's machine was built before or after #702 was merged.  @gforsyth, are you experiencing this issue on the current master?

**Comment by gforsyth** [comment]
Yes, @adqm, it's still an issue on current master.

**Comment by scopatz** [comment]
Hi all, I have put in #743, which should allow us to get past the silly `TypeError: unorderable types: NoneType() > int()` issue and see the full, real traceback.

**Comment by scopatz** [comment]
Or just magically make it work.

**Comment by gforsyth** [comment]
Would that there was still magic in this world, but progress(?)

If `/bin/bash` is the default shell, then it works

```console
rm -rI test
/usr/bin/rm: remote 1 argument recursively? 
```

If `xonsh` is the default shell and I log in on a TTY, then every `rm -rI test` gets backgrounded

```console
rm -rI test
/usr/bin/rm: remote 1 argument recursively? 
*[1] stopped: rm -rI test & (6829)
```

If `xonsh` is the default shell, I log in to a TTY and then `startx`, then the original failure behavior persists

```console
rm -rI test
/usr/bin/rm: remove 1 argument recursively? /usr/bin/rm: error closing file
```

No tracebacks present for any of these with `$XONSH_SHOW_TRACEBACK=True`.

**Comment by scopatz** [comment]
hmmm.  Just tried to reproduce this in mlterm on Ubuntu 15.10 with xonsh as the login shell and could not.

**Comment by scopatz** [comment]
Ahhhh!   I can reproduce it if `$XONSH_STORE_STDOUT = False`.  I bet this has something to do with flushing a `sys.stdout`

**Comment by scopatz** [comment]
OK, turning the trace on for `base_shell`, I see what is happening:

```console
scopatz@artemis ~ $ from xonsh import base_shell
scopatz@artemis ~ $ base_shell.__file__
'/home/scopatz/.local/lib/python3.5/site-packages/xonsh/base_shell.py'
scopatz@artemis ~ $ trace on /home/scopatz/.local/lib/python3.5/site-packages/xonsh/base_shell.py
.local/lib/python3.5/site-packages/xonsh/base_shell.py:151:            ts1 = time.time()
<snip>
scopatz@artemis ~ $ rm -rI test
<snip>
/bin/rm: remove 1 argument recursively? .local/lib/python3.5/site-packages/xonsh/base_shell.py:202:    def settitle(self):
.local/lib/python3.5/site-packages/xonsh/base_shell.py:204:        env = builtins.__xonsh_env__  # pylint: disable=no-member
.local/lib/python3.5/site-packages/xonsh/base_shell.py:205:        term = env.get('TERM', None)
.local/lib/python3.5/site-packages/xonsh/base_shell.py:208:        if term is None or term in ['dumb', 'eterm-color', 'linux']:
.local/lib/python3.5/site-packages/xonsh/base_shell.py:210:        t = env.get('TITLE')
.local/lib/python3.5/site-packages/xonsh/base_shell.py:211:        if t is None:
.local/lib/python3.5/site-packages/xonsh/base_shell.py:213:        t = format_prompt(t)
.local/lib/python3.5/site-packages/xonsh/base_shell.py:214:        if ON_WINDOWS and 'ANSICON' not in env:
.local/lib/python3.5/site-packages/xonsh/base_shell.py:218:            os.write(1, "\x1b]2;{0}\x07".format(t).encode())

*[1] stopped: rm -rI test & (13312)
.local/lib/python3.5/site-packages/xonsh/base_shell.py:151:            ts1 = time.time()
.local/lib/python3.5/site-packages/xonsh/base_shell.py:152:            if hist.last_cmd_rtn is None:
.local/lib/python3.5/site-packages/xonsh/base_shell.py:153:                hist.last_cmd_rtn = 0  # returncode for success
.local/lib/python3.5/site-packages/xonsh/base_shell.py:163:            ts1 = ts1 or time.time()
.local/lib/python3.5/site-packages/xonsh/base_shell.py:164:            self._append_history(inp=src, ts=[ts0, ts1], tee_out=tee.getvalue())
.local/lib/python3.5/site-packages/xonsh/base_shell.py:240:    def _append_history(self, tee_out=None, **info):
.local/lib/python3.5/site-packages/xonsh/base_shell.py:242:        _ = self
.local/lib/python3.5/site-packages/xonsh/base_shell.py:243:        hist = builtins.__xonsh_history__  # pylint: disable=no-member
.local/lib/python3.5/site-packages/xonsh/base_shell.py:244:        info['rtn'] = hist.last_cmd_rtn
.local/lib/python3.5/site-packages/xonsh/base_shell.py:245:        tee_out = tee_out or None
.local/lib/python3.5/site-packages/xonsh/base_shell.py:246:        last_out = hist.last_cmd_out or None
.local/lib/python3.5/site-packages/xonsh/base_shell.py:247:        if last_out is None and tee_out is None:
.local/lib/python3.5/site-packages/xonsh/base_shell.py:248:            pass
.local/lib/python3.5/site-packages/xonsh/base_shell.py:255:        hist.append(info)
.local/lib/python3.5/site-packages/xonsh/base_shell.py:256:        hist.last_cmd_rtn = hist.last_cmd_out = None
.local/lib/python3.5/site-packages/xonsh/base_shell.py:165:            tee.close()
.local/lib/python3.5/site-packages/xonsh/base_shell.py:166:        if builtins.__xonsh_exit__:  # pylint: disable=no-member
.local/lib/python3.5/site-packages/xonsh/base_shell.py:202:    def settitle(self):
.local/lib/python3.5/site-packages/xonsh/base_shell.py:204:        env = builtins.__xonsh_env__  # pylint: disable=no-member
.local/lib/python3.5/site-packages/xonsh/base_shell.py:205:        term = env.get('TERM', None)
.local/lib/python3.5/site-packages/xonsh/base_shell.py:208:        if term is None or term in ['dumb', 'eterm-color', 'linux']:
.local/lib/python3.5/site-packages/xonsh/base_shell.py:210:        t = env.get('TITLE')
.local/lib/python3.5/site-packages/xonsh/base_shell.py:211:        if t is None:
.local/lib/python3.5/site-packages/xonsh/base_shell.py:213:        t = format_prompt(t)
.local/lib/python3.5/site-packages/xonsh/base_shell.py:214:        if ON_WINDOWS and 'ANSICON' not in env:
.local/lib/python3.5/site-packages/xonsh/base_shell.py:218:            os.write(1, "\x1b]2;{0}\x07".format(t).encode())
```

Basically, what is going on is that the command (`rm` here) is run *before* and asynchronously to setting the title.  By when the interactive prompt pops up, the `settitle()` call then answers automatically for the user, which here happens to be false, thus quitting the command. Xonsh job control then detects that the process has been stopped, and reports it in a way that assumes the process was backgrounded, even though it was stopped internally.

**Comment by scopatz** [comment]
Basically, settitle() is answering the rm question for us.

**Comment by adqm** [comment]
That is so bizarre.

Just to add my experience to the pile: I always operate with `$XONSH_STORE_STDOUT = False`, and I haven't been able to reproduce.  On a base level, it doesn't make any sense to me how writing to the `xonsh` process's standard output could possibly be caught by the subprocess's standard input.  I am very confused.

I guess one way to test would be to set the title before we actually start running the commands, rather than after we have started them.  I wish I could reproduce so I could test myself.  Maybe I'll set up an Arch VM...

Also, the job control issue is concerning to me.  I've had similar issues recently (jobs reporting as backgrounded when they are actually still running, often right after launching them).  I don't remember this ever being an issue before, but I did take quite a long time away from working on xonsh, so I can try a git bisect to find when the particular issues I'm seeing started.  That said, it's a Heisenbug, so it might take some effort to track it down.

**Comment by ajstein** [comment]
I see @adqm is working on a branch, xoreutils, which looks very promising and interesting, as it pertains to the job backgrounding problem and general performance of some shell primitives.

In the interim, as I do not want to sit idly by, is there anything I can do to help further this? Debug output, tests, something?  I feel bad opening it, then standing on the sideline.

**Comment by scopatz** [comment]
@ajstein - if you could come up with a way to set the title *before* starting the subprocess, maybe by moving the settitle stuff into the job control(?) or something, that would be awesome.  Sorry I haven't had a chance to dig into this.

I would stress that this is a generic problem that even xoreutils won't fully fix, ie you can still go to `/bin/rm -rI` and experience this problem.

**Comment by ajstein** [comment]
Ok.  I am pretty newb at this stuff, so that should keep me busy all weekend!

**Comment by scopatz** [comment]
Yeah, this is a pretty nasty logic problem :).  The important thing to move around is where the `settitle()` gets called.

**Comment by ajstein** [comment]
So I was playing with this, and is it possible that changing the calling order here made a difference?

https://github.com/ajstein/xonsh/commit/8c78d18fdc45406400e2c7a74282c45081daf22c

That said, I am not sure why and I am by far not an expert.  In quick testing this resolved the issue.  I would not dare commit this, but thought this might help explain things.  Thoughts or am I way off track?

**Comment by adqm** [comment]
Hmmmmm it's not obvious to me why that change would fix the issue, but empirical evidence beats my intuition :)  If it works, I'm all for it.  I was never able to reproduce the issue, so I can't be any help with testing, I'm afraid.

I had thought that it was something with `settitle` being called after we had given control of the terminal to the child process (that's the only way it would make sense to me that the child process was grabbing from the shell's stdout), but if that were the issue, moving the code to this new spot wouldn't actually make a difference (since the code in between wasn't what was giving the terminal away.  Maybe it is a timing/race condition kind of thing?  But if that were the case, the thing we would be racing against is hitting the call for `wait_for_active_job`, which, because we're running synchronously, definitely shouldn't happen until after the title has been set.  I'm very confused.  :confused:

**Comment by scopatz** [comment]
It makes sense to me that this would "fix" the issue, because you are moving it *before* the `add_job()` call.  However, this probably breaks the actual content of the title, which uses the [`_current_job()` ](https://github.com/scopatz/xonsh/blob/master/xonsh/environ.py#L955) function to get the job name.  

I bet if you looked at the title for a long-running process like `sleep 2` the title would not be set correctly, or at least as it is set on master.  I think part of the problem stems from using job control to set the title.

**Comment by scopatz** [comment]
Note that one possibility is for job control itself to set the title *right* before it starts running a new job or right before switching to a new active job.

**Comment by ajstein** [comment]
Well it is not obvious to me either, @adqm.  Haha.  I was surprised.  I was not even looking at the title rendering.  I just wanted to understand while playing with `trace` on the `base_shell.py` like scoptaz demonstrated before.  I gave it a shot and wanted to see how the order of functions went.  I gave this a try.  I traditionally run `mlterm` and `emacs` full screen with multi-term.  So I do not see titles, as your linux and eterm `$TERM` cases indicate in the code.

I am just trying to understand the code better.  If i have a better suggestion later, I will let you know.  I just wanted to report back.
